lib/src/optimism/composition.rs

```diff
@@ -0,0 +1,272 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#[cfg(target_os = "zkvm")]
+use risc0_zkvm::{guest::env, serde::to_vec, sha::Digest};
+use serde::{Deserialize, Serialize};
+use zeth_primitives::{block::Header, tree::MerkleMountainRange, BlockHash, BlockNumber};
+
+use crate::optimism::DeriveOutput;
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct ComposeInput {
+    pub derive_image_id: [u32; 8],
+    pub compose_image_id: [u32; 8],
+    pub operation: ComposeInputOperation,
+    pub eth_chain_root: [u8; 32],
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub enum ComposeInputOperation {
+    PREP {
+        eth_blocks: Vec<Header>,
+        mountain_range: MerkleMountainRange,
+        prior: Option<ComposeOutput>,
+    },
+    LIFT {
+        derivation: DeriveOutput,
+        eth_tail_proof: MerkleMountainRange,
+    },
+    JOIN {
+        left: ComposeOutput,
+        right: ComposeOutput,
+    },
+    FINISH {
+        prep: ComposeOutput,
+        aggregate: ComposeOutput,
+    },
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, Eq, PartialEq)]
+pub struct ComposeOutput {
+    pub derive_image_id: [u32; 8],
+    pub compose_image_id: [u32; 8],
+    pub operation: ComposeOutputOperation,
+    pub eth_chain_root: [u8; 32],
+    pub eth_chain_root_validated: bool,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, Eq, PartialEq)]
+pub enum ComposeOutputOperation {
+    PREP {
+        eth_tail: BlockHash,
+    },
+    AGGREGATE {
+        op_head: (BlockNumber, BlockHash),
+        op_tail: (BlockNumber, BlockHash),
+        eth_tail: (BlockNumber, BlockHash),
+    },
+}
+
+impl ComposeInput {
+    pub fn process(self) -> ComposeOutput {
+        match self.operation {
+            ComposeInputOperation::PREP {
+                eth_blocks,
+                mountain_range,
+                prior,
+            } => {
+                // Check initial data
+                let (mut eth_tail, mut mountain_range) = if let Some(prior_output) = prior {
+                    #[cfg(target_os = "zkvm")]
+                    {
+                        // A valid receipt should be provided for prior aggregation
+                        let compose_journal = to_vec(&prior_output)
+                            .expect("Failed to encode prior aggregation journal");
+                        env::verify(Digest::from(self.compose_image_id), &compose_journal)
+                            .expect("Failed to validate prior aggregation");
+                    }
+                    // Validate context
+                    assert_eq!(self.derive_image_id, prior_output.derive_image_id);
+                    assert_eq!(self.compose_image_id, prior_output.compose_image_id);
+                    assert_eq!(self.eth_chain_root, prior_output.eth_chain_root);
+                    // Only append merkle range from preparation outputs
```

Comment:
`PREP` doesn't have to be done in one-shot, just in case the input size is too large and causes memory problems. You can build the Merkle tree incrementally by sequentially joining multiple `PREP`s. This comment just means that we'll continue the PREP only if `prior` is also a PREP, in which case `eth_chain_root_validated` is always true.